src/ngx_http_lua_log.c

```diff
@@ -313,4 +314,113 @@ ngx_http_lua_inject_log_consts(lua_State *L)
     /* }}} */
 }
 
+
+#ifndef NGX_LUA_NO_FFI_API
+int
+ngx_http_lua_ffi_errlog_set_filter_level(int level, u_char *err,
+    size_t *errlen)
+{
+#ifdef HAVE_INTERCEPT_ERROR_LOG_PATCH
+    ngx_http_lua_log_ringbuf_t     *ringbuf;
+
+    ringbuf = ngx_cycle->intercept_error_log_data;
+
+    if (ringbuf == NULL) {
+        *errlen = ngx_snprintf(err, *errlen,
+                               "API \"ngx.errlog_filter\" depends on directive "
+                               "\"lua_intercept_error_log\"")
+                  - err;
+        return NGX_ERROR;
+    }
+
+    if (level > NGX_LOG_DEBUG || level < NGX_LOG_STDERR) {
+        *errlen = ngx_snprintf(err, *errlen, "bad log level: %d", level)
+                  - err;
+        return NGX_ERROR;
+    }
+
+    ringbuf->filter_level = level;
+    return NGX_OK;
+#else
+    *errlen = ngx_snprintf(err, *errlen,
+                           "missing intercept error log patch in the nginx "
+                           "core")
+              - err;
+    return NGX_ERROR;
+#endif
+}
+
+
+int
+ngx_http_lua_ffi_errlog_count(u_char *err, size_t *errlen)
+{
+#ifdef HAVE_INTERCEPT_ERROR_LOG_PATCH
+    ngx_http_lua_log_ringbuf_t     *ringbuf;
+
+    ringbuf = ngx_cycle->intercept_error_log_data;
+
+    if (ringbuf == NULL) {
+        *errlen = ngx_snprintf(err, *errlen,
+                               "API \"ngx.errlog\" depends on directive "
+                               "\"lua_intercept_error_log\"")
+                  - err;
+        return NGX_ERROR;
+    }
+
+    return ringbuf->count;
+#else
+    *errlen = ngx_snprintf(err, *errlen,
+                           "missing intercept error log patch in the nginx "
+                           "core")
+              - err;
+    return NGX_ERROR;
+#endif
+}
+
+
+int
+ngx_http_lua_ffi_errlog(ngx_http_lua_ffi_table_elt_t *out, size_t max,
+    u_char *err, size_t *errlen)
+{
+#ifdef HAVE_INTERCEPT_ERROR_LOG_PATCH
+    void            *data = NULL;
+    size_t           len, i, count;
+    int              log_level;
+
+    ngx_http_lua_log_ringbuf_t     *ringbuf;
+
+    ringbuf = ngx_cycle->intercept_error_log_data;
+
+    if (ringbuf == NULL) {
+        *errlen = ngx_snprintf(err, *errlen,
+                               "API \"ngx.get_log\" depends on directive "
+                               "\"lua_intercept_error_log\"")
+                  - err;
+        return NGX_ERROR;
+    }
+
+    count = ringbuf->count;
+    if (count > max) {
+        count = max;
+    }
+
+    for (i = 0; i < count; i++) {
+        ngx_http_lua_log_ringbuf_read(ringbuf, &log_level, &data, &len);
+        out[i].key.len = log_level;
+        out[i].value.len = len;
+        out[i].value.data = data;
+    }
+
+    return NGX_OK;
```

Comment:
I think it's better for this function to return the `count` directly so that the caller does not have to call `ngx_http_lua_ffi_get_errlog_count` first. This can be more efficient.